alf/bin/grid_search.py

```diff
@@ -0,0 +1,159 @@
+                    'Path to the grid search config file.')
+FLAGS = flags.FLAGS
+r"""Grid search.
+
+To run grid search on ddpg for gym Pendulum:
+```bash
+cd ${PROJECT}/alf/examples;
+python -m alf.bin.grid_search \
+  --root_dir=~/tmp/ddpg_pendulum \
+  --search_config=ddpg_grid_search.json \
+  --gin_file=ddpg_pendulum.gin \
+  --gin_param='create_environment.num_parallel_environments=8' \
+  --alsologtostderr
+```
+"""
+
+
+class GridSearchConfig(object):
+    """ Grid Search Config
+
+    Grid search config file should be json format
+    For example:
+    {
+        "desc": "desc text",
+        "max_worker_num": 8,
+        "parameters": {
+            "create_ddpg_algorithm.actor_learning_rate": [1e-3, 8e-4],
+            "OneStepTDLoss.gamma":"(0.995, 0.99)",
+            "param3_name": param3_value,
+              ...
+        }
+        ...
+    }
+    `max_worker_num` is max number of parallel search worker processes
+    `parameters` is a dict(param_name=param_value,) of configured search space .
+    param_name is a gin configurable argument str and param_value must be an
+    iterable python object or a str that can be evaluated to an iterable object.
+
+    See `ddpg_grid_search.json` for an example.
+    """
+
+    def __init__(self, conf_file):
+        """
+        Args:
+            conf_file (str): Path to the config file.
+        """
+        with open(conf_file) as f:
+            param_keys = []
+            param_values = []
+            conf = json.loads(f.read())
+            for key, value in conf['parameters'].items():
+                if isinstance(value, str):
+                    value = eval(value)
+                param_keys.append(key)
+                param_values.append(value)
+
+        self._desc = conf['desc']
+        self._param_keys = param_keys
+        self._param_values = param_values
+        self._max_worker_num = conf['max_worker_num']
+
+    @property
+    def desc(self):
+        return self._desc
+
+    @property
+    def param_keys(self):
+        return self._param_keys
+
+    @property
+    def param_values(self):
+        return self._param_values
+
+    @property
+    def max_worker_num(self):
+        return self._max_worker_num
+
+
+class GridSearch(object):
+    """Grid Search"""
+
+    def __init__(self, conf_file):
+        """Create GridSearch instance
+        Args:
+            conf_file (str): Path to the config file.
+        """
+        self._conf = GridSearchConfig(conf_file)
+
+    def run(self):
+        """Run trainings with all possible parameter combinations in configured space
+        """
+
+        processes = []
+        param_keys = self._conf.param_keys
+        param_values = self._conf.param_values
+        max_worker_num = self._conf.max_worker_num
+
+        # `worker_queue` is just a status queue to control the number for working process.
+        # And to simplify, process pool is not used here, just spawn process for new
```

Comment:
Sorry for this late review. Why not just use Pool here? The code will be much shorter and less error-prone in that way.

Reply:
Yeah , u are right , code will be much shorter  and easy to understand here . But some reset  work needs to be done before job start  when multi training taken on the same process , such as reset `global_step`  etc, and it may be a problem to make sure all global states are reset. 
so ,  using  processes to isolate jobs is better here .

Reply:
OK, I see.

Reply:
I found out that you can specify an argument `maxtasksperchild=1` to process.Pool to have the same behavior. 

"maxtasksperchild is the number of tasks a worker process can complete before it will exit and be replaced with a fresh worker process, to enable unused resources to be freed. The default maxtasksperchild is None, which means worker processes will live as long as the pool."

Reply:
so  , Implementation is equivalet